ex2_hal/iris/hardware_interface/source/iris.c

```diff
@@ -0,0 +1,460 @@
+/*
+ * Copyright (C) 2015  University of Alberta
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include "FreeRTOS.h"
+#include <string.h>
+#include <stdlib.h>
+
+#include "iris.h"
+#include "iris_spi.h"
+#include "logger.h"
+
+/*
+ * Optimization points
+ * - Full error coverage is desired
+ */
+
+
+/**
+ * @brief
+ *   Sends take a picture command to Iris
+ *
+ * @return
+ *   Returns IRIS_HAL_OK if equipment handler returns IRIS_ACK, else IRIS_HAL_ERROR
+ **/
+IrisHALReturn iris_take_pic() {
+    IrisLowLevelReturn ret;
+
+    controller_state = SEND_COMMAND;
+
+    while (controller_state != FINISH && controller_state != ERROR_STATE) {
+        switch (controller_state) {
+            case SEND_COMMAND:
+            {
+                ret = send_command(IRIS_TAKE_PIC);
+                if (ret == IRIS_ACK) {
+                    controller_state = FINISH;
+                } else {
+                    controller_state = ERROR_STATE;
+                }
+                break;
+            }
+            case FINISH:
+            {
+                ex2_log("Iris returns ACK on take a picture command");
+                return IRIS_HAL_OK;
+            }
+            case ERROR_STATE:
+            {
+                ex2_log("Iris returns NACK on take a picture command");
+                return IRIS_HAL_ERROR;
+            }
+        }
+    }
+    return IRIS_HAL_ERROR;
+}
+
+/**
+ * @brief
+ *   Sends a get image length command to Iris, and expects to receive the
+ *   image length
+ *
+ * @return
+ *   Returns the length in bytes of image stored on Iris
+ **/
+IrisHALReturn iris_get_image_length(uint32_t *image_length) {
```

Comment:
How do we decide which image we're getting the size of?

Reply:
As of now we do not have a explicit on the Iris side to get image length. I assume that we will be getting the image length from the NAND flash